public/app/features/live/live.ts

```diff
@@ -113,51 +164,32 @@ class CentrifugeSrv implements GrafanaLiveSrv {
     return this.connectionState.asObservable();
   }
 
-  initChannel<T>(path: string, handler: ChannelHandler<T>) {
-    if (this.channels[path]) {
-      console.log('Already connected to:', path);
-      return;
-    }
-    const c: Channel = {
-      subject: new Subject<T>(),
-    };
-    this.channels[path] = c;
-
-    console.log('initChannel', this.centrifuge.isConnected(), path, handler);
-    const callbacks: SubscriptionEvents = {
-      ...this.standardCallbacks,
-      publish: (ctx: PublicationContext) => {
-        // console.log('GOT', JSON.stringify(ctx.data), ctx);
-        const v = handler.onPublish(ctx.data);
-        c.subject.next(v);
-      },
-    };
-    c.subscription = this.centrifuge.subscribe(path, callbacks);
-  }
-
-  getChannelStream<T>(path: string): Observable<T> {
-    let c = this.channels[path];
-    if (!c) {
-      this.initChannel(path, noopChannelHandler);
-      c = this.channels[path];
-    }
-    return c!.subject.asObservable();
+  getChannelStream<T>(plugin: string, path: string): Observable<T> {
+    return from(this.getChannel<T>(plugin, path)).pipe(
+      take(1), // This converts the promise to observable
+      mergeMap(c => c.stream)
+    );
   }
 
   /**
    * Send data to a channel.  This feature is disabled for most channels and will return an error
    */
-  publish<T>(channel: string, data: any): Promise<T> {
-    return this.centrifuge.publish(channel, data);
+  async publish<T>(plugin: string, path: string, data: any): Promise<T> {
+    const support = await this.getChannelSupport(plugin);
+    const handler = support.getChannelHandler(path);
+    if (!handler) {
+      throw new Error(`Invalid path: ${path}`);
+    }
+    if (!handler.onPublish) {
+      throw new Error(`Channel ${path} does not allow publishing`);
+    }
+    data = handler.onPublish(data);
+
+    // Writes a message over the websocket to grafana server
+    return this.centrifuge.publish(`${plugin}/${path}`, data);
```

Comment:
not sure I understand the handler.onPublish and onMessageReceived they are both event pre-processors?  what would the use case for handler.onPublish be? and handler. onMessageReceived

Reply:
`onMessageReceived` could get a single data point and will put it into a circular DataFrame -- the observable you subscribe to will give you a DataFrame -- or whatever makes sense for the channel.

The `onPublish` one is more awkward and I am not totally sure how it should work :)  We need some way for the channel handler to intercept and validate messages before we send them or even if we send them.  Currently the backend allows anyone to write messages to any channel -- that should not be the answer long term, but makes it easy to write cool features entirely in the frontend websocket.